### PR TITLE
Fixes #33451 - use upsert all for content indexing

### PR DESCRIPTION
**Pull request by jlsherrill**
#### What are the changes introduced in this pull request?
Rewrites all of content indexing to use bulk actions instead of single AR creates.  This does two things:
1) avoids the use of the 'active record retry' method that just catches conflict errors (from concurrent imports), and retries.  This was adding errors to the postgresql logs
2) makes the process much much faster.  In my timings, indexing rhel7 os went from 800 seconds to 126 seconds.   Reindex went from 221 seconds to 140 seconds.

#### Considerations taken when implementing this change?
This pr brought to light a bunch of missing uniqueness constraints in the database that need correcting.

#### What are the testing steps for this pull request?

We need to test all types of content syncing and cv publish with & without filters.  You would also want to make sure no data is missing on each row (i.e. errata description).  So examining the db tables/models of:

Katello::Rpm
Katello::Erratum
Katello::PackageGroup
Katello::ModuleStream

Katello::Deb

Katello::DockerTag
Katello::DockerManifest
Katello::DockerManifestList

Katello::GenericContentUnit  (python & ostree)

Katello::FileUnit

would be good.

Need to verify that this does not break https://github.com/Katello/katello/pull/9854

### CODE REVIEW

File: db/migrate/20211220185935_clean_duplicate_content_units.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class CleanDuplicateContentUnits < ActiveRecord::Migration[6.0]
+  def up
+    #TODO: HANDLE DUPLICATE CLENAUP
```
Comment (jlsherrill) [inline db/migrate/20211220185935_clean_duplicate_content_units.rb:3]: TODO: cleanup duplicate data, i think everything else can be tested atm

**Comment by jlsherrill** [comment]
[test katello]

File: db/migrate/20211220185935_clean_duplicate_content_units.rb [hunk 2]
```diff
@@ -0,0 +1,144 @@
+class CleanDuplicateContentUnits < ActiveRecord::Migration[6.0]
+  def find_duplicate_groups(model, fields)
+    model.having("count(*) > 1").group(*fields).pluck("string_agg(id::TEXT, ',') ").map do |ids|
```
Comment (parthaa) [inline db/migrate/20211220185935_clean_duplicate_content_units.rb:3]: Interesting. Not seen string_agg before. Guessing it groups ids separated by comma.

**Review by parthaa — APPROVED** [review]
Ok I tested the following
1) Sync & Migrating  App Stream
2) Sync & Migrating RHEL 6 kickstart
3) Docker Fedora SSH
4) Published with and Without filters. 

Examined the Rpms/Errata/ModuleStreams and Docker stuff. Seemed to work well.
Works well for me.

@jlsherrill  Feel free to get a second opinion on the coding aspects :). LGTM

**Comment by jlsherrill** [comment]
[test katello]

**Comment by ianballou** [comment]
Was updating errata tested? Asking since I saw #9854 mentioned.

File: app/services/katello/content_unit_indexer.rb [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        index.columns.sort == columns.sort
+      end
+      if found
+        found.columns
+      else
+        fail "Unable to find unique index for #{columns} on table #{self.repository_association_class.table_name}"
+      end
+    end
+
+    def copy_repository_associations(source_repo, dest_repo)
```
Comment (ianballou) [inline app/services/katello/content_unit_indexer.rb:191]: Is this being used here? I see it in `pulp_database_unit.rb` as well.

File: app/services/katello/content_unit_indexer.rb [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+      @content_type = content_type
+      @pulp_content_ids = pulp_content_ids
+      @optimized = optimized
+      @other_time = 0
+      @import_time = 0
+      @assoc_time = 0
+    end
+
+    def report
+      Rails.logger.error("Times: #{@other_time}, #{@import_time}, #{@assoc_time}")
```
Comment (ianballou) [inline app/services/katello/content_unit_indexer.rb:17]: Is this an `error` just for debugging?

File: app/services/katello/content_unit_indexer.rb [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+          import_time { @model_class.upsert_all(to_insert, unique_by: :pulp_id) }
+        else
+          import_time { @model_class.insert_all(to_insert, unique_by: :pulp_id) }
+        end
+
+        import_associations(units) if @repository
+      end
+
+      if @model_class.many_repository_associations && @repository
+        assoc_time { sync_repository_associations(association_tracker) } #:pulp_ids => ids_to_associate, :additive => true) }
```
Comment (ianballou) [inline app/services/katello/content_unit_indexer.rb:78]: What's going on with the comment?

Comment (jlsherrill) [reply app/services/katello/content_unit_indexer.rb:78]: oh i need to clean tha tup!

File: db/migrate/20211220185935_clean_duplicate_content_units.rb [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      ids.split(',').map(&:to_i)
+    end
+  end
+
+  def reassign_duplicate_associations(associated_models, duplicates, new_id, field, _unique_fields)
```
Comment (ianballou) [inline db/migrate/20211220185935_clean_duplicate_content_units.rb:8]: Is `_unique_fields` needed?

Comment (jlsherrill) [reply db/migrate/20211220185935_clean_duplicate_content_units.rb:8]: nope!  good catch

**Comment by ianballou** [comment]
I got an error when indexing errata from the EPEL 7 x86 repository. I think the issue is the title.  One of the errata (FEDORA-EPEL-2017-96b04b2732) has a title longer than 255 characters.

Error on the errata table:
```
ActiveRecord::ValueTooLong: PG::StringDataRightTruncation: ERROR:  value too long for type character varying(255)
```

**Comment by ianballou** [comment]
It must be because `upsert_all` doesn't trigger callbacks or validations.  I'm guessing this just isn't running: https://github.com/Katello/katello/blob/master/app/models/katello/erratum.rb#L50

**Comment by ianballou** [comment]
I just hit an error with indexing file units. I added all my repositories to a CV, made an erratum and an RPM filter, then published.  This is the error during indexing of the file repo:

```
16:01:18 rails.1   | 2022-02-04T16:01:18 [E|bac|27287d6c] no implicit conversion of nil into String (TypeError)
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/pulp3/file_unit.rb:23:in `basename'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/pulp3/file_unit.rb:23:in `generate_model_row'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/content_unit_indexer.rb:34:in `block (2 levels) in import_all'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/content_unit_indexer.rb:28:in `map'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/content_unit_indexer.rb:28:in `block in import_all'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/pulp3/pulp_content_unit.rb:97:in `yield'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/pulp3/pulp_content_unit.rb:97:in `block (2 levels) in pulp_units_batch_for_repo'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/pulp3/pulp_content_unit.rb:85:in `loop'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/pulp3/pulp_content_unit.rb:85:in `block in pulp_units_batch_for_repo'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/content_unit_indexer.rb:27:in `each'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/services/katello/content_unit_indexer.rb:27:in `import_all'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/models/katello/repository.rb:913:in `block (2 levels) in index_content'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/lib/katello/logging.rb:8:in `time'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/models/katello/repository.rb:912:in `block in index_content'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/models/katello/repository.rb:911:in `each'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/models/katello/repository.rb:911:in `index_content'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/lib/actions/katello/repository/index_content.rb:22:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:582:in `block (3 levels) in execute_run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/rails_executor_wrap.rb:14:in `block in run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/activesupport-6.0.3.7/lib/active_support/execution_wrapper.rb:88:in `wrap'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/rails_executor_wrap.rb:13:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action/progress.rb:31:in `with_progress_calculation'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action/progress.rb:17:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:11:in `block in run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:22:in `with_locale'
16:01:18 rails.1   |  27287d6c | /home/vagrant/katello/app/lib/actions/middleware/keep_locale.rb:11:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/load_setting_values.rb:20:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_request_id.rb:15:in `block in run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_request_id.rb:52:in `restore_current_request_id'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_request_id.rb:15:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_timezone.rb:15:in `block in run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_timezone.rb:44:in `restore_curent_timezone'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_timezone.rb:15:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_taxonomies.rb:15:in `block in run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_taxonomies.rb:45:in `restore_current_taxonomies'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_taxonomies.rb:15:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:32:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:27:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware.rb:19:in `pass'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_user.rb:15:in `block in run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_user.rb:54:in `restore_curent_user'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/bundler/gems/foreman-tasks-831bd3fe186c/app/lib/actions/middleware/keep_current_user.rb:15:in `run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/stack.rb:23:in `call'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/middleware/world.rb:31:in `execute'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:581:in `block (2 levels) in execute_run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:580:in `catch'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:580:in `block in execute_run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:483:in `block in with_error_handling'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:483:in `catch'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:483:in `with_error_handling'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:575:in `execute_run'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/action.rb:296:in `execute'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:18:in `block (2 levels) in execute'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/execution_plan/steps/abstract.rb:167:in `with_meta_calculation'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:17:in `block in execute'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:32:in `open_action'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:16:in `execute'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/director.rb:69:in `execute'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/executors/parallel/worker.rb:15:in `block in on_message'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/executors.rb:18:in `run_user_code'
16:01:18 rails.1   |  27287d6c | /home/vagrant/foreman/.vendor/ruby/2.7.0/gems/dynflow-1.6.3/lib/dynflow/executors/parallel/worker.rb:14:in `on_message'
16:01:18 rails.1   |  27287d6c | [ concurrent-ruby ]
```

**Comment by ianballou** [comment]
You can reproduce the issue by making an RPM filter in a CV  that only has a file repository.

**Review by ianballou — APPROVED** [review]
Looks all good now!

**Pull request merged**